### ISSUE TITLE
Possible inconsistent behavior

### ISSUE THREAD

**Issue by pierreganty** [vim/vim#8346]
**Describe the bug**

After entering select mode one can switch back and forth to visual mode using `<C-g>`.
It also works the same with `<C-o>` for different reasons suggested in `visual.txt`
```
Commands in Select mode:
[...]
- CTRL-O switches to Visual mode for the duration of one command. *v_CTRL-O*
```
and
```
Commands (starting with) ".", "&", CTRL-^, "Z", CTRL-], CTRL-T, CTRL-R, CTRL-I
and CTRL-O cause a beep and Visual mode continues.
```
When you extract the mode using maps, the behavior between `<C-o>` and `<C-g>` slightly differs and it is unclear why. 

**To Reproduce**
Detailed steps to reproduce the behavior:
1. Run `vim -Nu NONE -S mwe.vim9` where `mwe.vim9` is
```viml
vim9script

snoremap <expr> s- execute('g:mode = mode(1)')->slice(0, 0) .. '<Ignore>'
xnoremap <expr> X- execute('g:mode = mode(1)')->slice(0, 0) .. '<Ignore>'

feedkeys('ghs-', 'x')
assert_equal('s', g:mode)

unlet g:mode
feedkeys("\<C-o>X-", 'x')
assert_equal('v', g:mode)

unlet g:mode
feedkeys("\<C-o>s-", 'x')
assert_equal('s', g:mode)

echom v:errors
```
2. It fails at the last assert because `g:mode` is undefined.
3. Now if you split the last `feedkeys` into first `<C-o>` then `s-` as follows:
```viml
feedkeys("\<C-o>", 'x')
feedkeys("s-", 'x')
````
you obtain a script that does not fail.

**Expected behavior**
I would expect the last `feedkeys("\<C-o>s-", 'x')`  and `feedkeys("\<C-o>", 'x') feedkeys("s-", 'x')` to behave the same.

**Environment (please complete the following information):**
 - Vim version  8.2.2932 
 - macOS 11.3.1
 - Terminal.app 

**Additional context**
Changing `<C-o>` for `<C-g>` does not yield the same problem: `feedkeys("\<C-g>s-", 'x')` works fine.

**Comment by pierreganty** [comment]
I suspect that it might have a relationship with the following issue https://github.com/neovim/pynvim/issues/484

**Comment by lacygoill** [comment]
> I would expect the last feedkeys("\<C-o>s-", 'x') and feedkeys("\<C-o>", 'x') feedkeys("s-", 'x') to behave the same.

But they are not the same.  Where you break the sequence matters.  Just like where you put a comma in a sentence matters:

   > Let's eat, Grandma.

Versus:

   > Let's eat Grandma.

Regarding the failing `feedkeys("\<C-o>s-", 'x')`, pass the `t` flag so that Vim handles the keys as if they were typed interactively (and not coming from the expansion of a mapping):

    feedkeys("\<C-o>s-", 'xt')
                           ^

It should suppress the error:
```vim
vim9script
snoremap <expr> s- execute('g:mode = mode(1)')->slice(0, 0) .. '<Ignore>'
xnoremap <expr> X- execute('g:mode = mode(1)')->slice(0, 0) .. '<Ignore>'

feedkeys('ghs-', 'x')
assert_equal('s', g:mode)

unlet g:mode
feedkeys("\<C-o>X-", 'x')
assert_equal('v', g:mode)

unlet g:mode
feedkeys("\<C-o>s-", 'xt')
assert_equal('s', g:mode)

echom v:errors
```

**Comment by brammool** [comment]
Not sure what you are trying to do, but the help says:
- CTRL-O switches to Visual mode for the duration of one command. *v_CTRL-O*
At the end of "feedkeys(..., 'x')" the CTRL-O effect will be lost, thus it works differently.

**Comment by pierreganty** [comment]
@brammool I am trying to build, using automated techniques, an automata based model of vim.
(This is how I uncovered https://github.com/vim/vim/issues/8323)
Using these techniques I could get partial models. Below is an inferred automata (i.e. a Moore machine) for alphabet `'<Esc>', '<C-o>', 'gh', 'v', 'i'`.

Despite your (and @lacygoill ) precious help, I haven't yet figured out why `<C-o>` connects `s7` and `s9` back and forth but the same does not hold for `s1` and `s4`. I suspect my understanding of `feedkeys` is incorrect and I was not lucky with the various `feedkeys` mode I've tried.

![model](https://user-images.githubusercontent.com/2111160/121355845-9926c480-c930-11eb-8d48-e598566630e9.png)
